plugins/x86/x86_mov_offset.ml

```diff
@@ -0,0 +1,189 @@
+  end
+
+  module Mov_ao_ia32 = struct
+    type t = MOV8ao8 | MOV16ao16 | MOV32ao32
+    [@@deriving bin_io, sexp, compare, enumerate]
+
+    let asm_of_t t = match t with
+      | MOV8ao8   -> `AL      
+      | MOV16ao16 -> `AX
+      | MOV32ao32 -> `EAX
+
+    let semantic = Reg_to_mem
+  end
+
+  module Mov_ao_amd64 = struct
+    type t = MOV64ao8 | MOV64ao16 | MOV64ao32 | MOV64ao64
+    [@@deriving bin_io, sexp, compare, enumerate]
+
+    let asm_of_t t = match t with
+      | MOV64ao8  -> `AL
+      | MOV64ao16 -> `AX
+      | MOV64ao32 -> `EAX      
+      | MOV64ao64 -> `RAX
+
+    let semantic = Reg_to_mem
+  end
+end
+
+module Ver_38 = struct
+  module Mov_oa_ia32 = struct
+    type t =
+      | MOV8o16a
+      | MOV8o32a
+      | MOV16o32a
+      | MOV32o32a
+      | MOV16o16a
+      | MOV32o16a
+      | MOV64o32a
+    [@@deriving bin_io, sexp, compare, enumerate]
+
+    let asm_of_t op = match op with
+      | MOV8o16a  | MOV8o32a  -> `AL
+      | MOV16o16a | MOV16o32a -> `AX
+      | MOV32o16a | MOV32o32a -> `EAX
+      | MOV64o32a -> `EAX (** llvm claims it's RAX in 32-address mode *)
+        
+    let semantic = Reg_to_mem
+  end
+
+  module Mov_oa_amd64 = struct
+    type t = MOV8o64a | MOV16o64a | MOV32o64a | MOV64o64a
+    [@@deriving bin_io, sexp, compare, enumerate]
+
+    let asm_of_t op = match op with
+      | MOV8o64a  -> `AL
+      | MOV16o64a -> `AX
+      | MOV32o64a -> `EAX
+      | MOV64o64a -> `RAX
+
+    let semantic = Reg_to_mem
+  end
+
+  module Mov_ao_ia32 = struct
+    type t =
+      | MOV8ao16
+      | MOV8ao32
+      | MOV16ao32
+      | MOV32ao32
+      | MOV16ao16
+      | MOV32ao16
+      | MOV64ao32
+    [@@deriving bin_io, sexp, compare, enumerate]
+
+    let asm_of_t op = match op with
+      | MOV8ao16  | MOV8ao32  -> `AL
+      | MOV16ao16 | MOV16ao32 -> `AX
+      | MOV32ao16 | MOV32ao32 -> `EAX
+      | MOV64ao32 -> `EAX  (** llvm claims it's RAX in 32-address mode *)
+        
+    let semantic = Mem_to_reg
+  end
+
+  module Mov_ao_amd64 = struct
+    type t = MOV8ao64 | MOV16ao64 | MOV32ao64 | MOV64ao64
+    [@@deriving bin_io, sexp, compare, enumerate]
+
+    let asm_of_t op = match op with
+      | MOV8ao64  -> `AL
+      | MOV16ao64 -> `AX
+      | MOV32ao64 -> `EAX
+      | MOV64ao64 -> `RAX
+
+    let semantic = Mem_to_reg
+  end
+end
+
+module Make(V : Version) = struct
+  module IA32 = X86_backend.IA32
+  module AMD64 = X86_backend.AMD64
+  module Sema32 = Insn_semantic(X86_tools.IA32)
+  module Sema64 = Insn_semantic(X86_tools.AMD64)
+  open V
+
+  let register' insn back sema =
+    let module L = (val insn : T) in
+    let module B = (val back : X86_backend.T) in
+    let module S = (val sema : Semantic) in
+    List.iter L.all (fun op ->        
+        let f = S.lift (L.asm_of_t op) L.semantic in
+        let s = L.sexp_of_t op |> Sexp.to_string in
+        B.register s X86_opcode.Mov f)
+
+  let register () =
+    register' (module Mov_oa_ia32)  (module IA32)  (module Sema32);
```

Comment:
it is quite obvious, that `register` is not recursive, so you can just use `register` here without the ugly `'`. Alternatively, you can use another name, e.g., `add`